Static Allocator - Aligned allocations

emersonknapp opened the pull request:
* Remove unwanted `main` symbol
* Centralize copy-pasted code for getting static allocator
* Align the memory pool and returned memory from in the static allocator to std::max_align_t in order to make it all safe for optimized builds

Fixes osrf/osrf_testing_tools_cpp#30
Unblocks ros2/rmw_fastrtps#272

Signed-off-by: Emerson Knapp <eknapp@amazon.com>

emersonknapp commented:
@thomas-moulard Can you please run CI for this PR? Unfortunately we cannot restrict the test or build packages very much.

Gist: https://gist.githubusercontent.com/emersonknapp/c3646b3d48b6ad07b1c9144eac98dff7/raw/fdc4bae8a7a945f689c6748cf6fc66187b61df5c/ros2.repos
Additional BUILD args: --packages-up-to test_rclcpp test_rclpy test_cli test_cli_remapping rclpy
Additional TEST args: --packages-up-to test_rclcpp test_rclpy test_cli test_cli_remapping rclpy
CI Job: ci_launcher

thomas-moulard commented on osrf_testing_tools_cpp/src/memory_tools/impl/linux.cpp:
```diff
@@ -42,13 +42,21 @@ find_original_function(const char * name)
 
+// An amount of memory that is greater than what is needed for static initialization
+// for any test we run. It was found experimentally on Ubuntu Linux 16.04 x86_64.
+static const size_t STATIC_ALLOCATOR_SIZE = 0x800000;
```
constexpr?

thomas-moulard commented on osrf_testing_tools_cpp/src/memory_tools/impl/linux.cpp:
```diff
@@ -42,13 +42,21 @@ find_original_function(const char * name)
+
+// Contains global allocator to make 100% sure to avoid Static Initialization Order Fiasco.
+// "Construct on first use" idiom
+StaticAllocatorT * g_static_allocator() {
```
`g_static_allocator` => `allocate_or_get_static_allocator`

thomas-moulard commented on osrf_testing_tools_cpp/src/memory_tools/impl/static_allocator.hpp:
```diff
@@ -45,9 +45,10 @@ class StaticAllocator
   {
-    if (size <= size_t(std::distance(stack_pointer_, end_))) {
+    auto const aligned_size = align_up(size);
+    if (aligned_size <= static_cast<decltype(aligned_size)>(std::distance(end_, stack_pointer_))) {
```
why `decltype` here? We know it's `size_t`, no?

thomas-moulard commented on osrf_testing_tools_cpp/src/memory_tools/impl/static_allocator.hpp:
```diff
@@ -45,9 +45,10 @@ class StaticAllocator
   allocate(size_t size)
   {
-    if (size <= size_t(std::distance(stack_pointer_, end_))) {
+    auto const aligned_size = align_up(size);
```
`const auto` ?

thomas-moulard commented on osrf_testing_tools_cpp/src/memory_tools/impl/static_allocator.hpp:
```diff
@@ -105,7 +106,14 @@ class StaticAllocator
+    return (n+(alignment-1)) & ~(alignment-1);
+  }
+
+  static auto constexpr alignment = alignof(std::max_align_t);
```
`MAX_ALIGN` ? This needs a comment :)

emersonknapp replied:
moved to constant in namespace

thomas-moulard commented on osrf_testing_tools_cpp/src/memory_tools/impl/static_allocator.hpp:
```diff
@@ -105,7 +106,14 @@ class StaticAllocator
+  }
+
+  static auto constexpr alignment = alignof(std::max_align_t);
+  alignas(alignment) uint8_t memory_pool_[MemoryPoolSize];
```
This probably needs a comment too ;)

thomas-moulard commented on osrf_testing_tools_cpp/src/memory_tools/impl/static_allocator.hpp:
```diff
@@ -105,7 +106,14 @@ class StaticAllocator
+  static std::size_t
+  align_up(std::size_t n) noexcept
+  {
+    return (n+(alignment-1)) & ~(alignment-1);
```
Let's make this an inline function so that we can write tests for that this one. Would be nice to make it more readable too, or at least add comments.

emersonknapp replied:
This trick is used in Boost and shows up all over the place - I don't know that a unit test is really necessary? https://github.com/boostorg/align/blob/develop/include/boost/align/align_up.hpp#L19

thomas-moulard replied:
1. Tests are not there only to make sure your code works. It is there to ensure that the next person can understand it (tests are a type of documentation), modify it easily and trust it.
2. Reading your code, `std::max_align_t` is implementation defined meaning that it may or may not be signed type and may be an integral type of varying length... unlike the original implementation where `alignment` is always a `std::size_t`. Knowing that we will use this code on various architectures with various STL implementations, why take the risk of having this silently start returning bad results?

emersonknapp replied:
I see your points, I will add to the unit tests.

To elaborate specifically on point 2 though, I am only using `alignof(std::max_align_t)` which the alignof operator does return a `std::size_t` according to the spec, so the type `std::max_align_t` doesn't matter https://en.cppreference.com/w/cpp/language/alignof

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7063)](http://ci.ros2.org/job/ci_linux/7063/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3291)](http://ci.ros2.org/job/ci_linux-aarch64/3291/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5787)](http://ci.ros2.org/job/ci_osx/5787/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6879)](http://ci.ros2.org/job/ci_windows/6879/)

emersonknapp commented:
Any thoughts on how to unit test this? I am not able to reproduce the observed problem in a toy example - only in the context of a specific system dynamically linking against OpenSSL, using an optimized build of glibc6

emersonknapp commented:
@thomas-moulard The build failed because my arguments were wrong - the test and build args should have been:

--packages-up-to test_rclcpp test_communication test_cli test_cli_remapping rclpy

thomas-moulard commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7065)](http://ci.ros2.org/job/ci_linux/7065/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3293)](http://ci.ros2.org/job/ci_linux-aarch64/3293/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5789)](http://ci.ros2.org/job/ci_osx/5789/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6881)](http://ci.ros2.org/job/ci_windows/6881/)
Triggering a new build of ci_linux

thomas-moulard commented:
What about instantiating a static allocator in the unit test and checking the addresses are actually aligned?
Obviously a repro-case would be best, but it will be hard to write. I think we can find a trade-off there between repro-case and no tests.

wjwwood commented on osrf_testing_tools_cpp/src/memory_tools/impl/linux.cpp:
```diff
@@ -42,13 +42,21 @@ find_original_function(const char * name)
+
+// Contains global allocator to make 100% sure to avoid Static Initialization Order Fiasco.
+// "Construct on first use" idiom
+StaticAllocatorT * get_static_allocator() {
```
This package cannot depend on ament_lint, but if it could I would do it :) To that end please format this with the `{` on a new line and for my preference the return type on its own line:

```c++
StaticAllocatorT *
get_static_allocator()
{
```

Also, please make it `static`, as no one else needs to see it outside of the TU.

emersonknapp replied:
Why can't it? Does it introduce a circular dependency? Barring that, could we hook it up to a Travis job that runs linters non-intrusively?

Will address the formatting

thomas-moulard replied:
You could use `uncrustify --reformat` BTW even if you don't have the linter plugged into this package for this PR.

wjwwood replied:
I intentionally made this package pure cmake, rather than depend on ament so that some others could use it (it's only dependency is cmake).

We could have a travis job that installs and runs `ament_*` linters, but I just haven't done that.

wjwwood commented:
In general, looks good to me. It might fix aarch64 too, since we've had alignment issue on ARM in the past. Thanks for looking into it, I know the code is hairy :x

emersonknapp commented:
@wjwwood  I've addressed the formatting concerns and added in the statics. I'm looking at test_memory_tools.cpp and trying to figure how to add a unit test for static_allocator.hpp, which is buried in the source directory. Any ideas on that without having to restructure the project?

Other than that, it looks like the test failures on CI were `test_cli.test_params_yaml` timeout - which I am getting locally with or without this chage (btw do you know if there is an open issue for that? if not should I open it on system_tests? i think it's one of those things we keep seeing)

emersonknapp commented:
I've just pushed a commit that is an attempt at exposing the src dir to the tests without restructuring things - lmk if this is ok or if it's a kludge - I wouldn't consider myself enough of a CMake expert to determine.

jacobperron commented:
> Other than that, it looks like the test failures on CI were test_cli.test_params_yaml timeout - which I am getting locally with or without this chage (btw do you know if there is an open issue for that? if not should I open it on system_tests? i think it's one of those things we keep seeing)

Here's the issue for the known flake https://github.com/ros2/build_cop/issues/166

dirk-thomas commented:
Testing `rclcpp` and `test_rclcpp`:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7078)](http://ci.ros2.org/job/ci_linux/7078/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3308)](http://ci.ros2.org/job/ci_linux-aarch64/3308/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5803)](http://ci.ros2.org/job/ci_osx/5803/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6896)](http://ci.ros2.org/job/ci_windows/6896/)

emersonknapp commented:
Regarding time source test failures in above - I am able to reproduce locally on master ros2.repos without this patch.

As for this patch, I believe the only packages this really affects are `rcl` and `rcutils`, because from my understanding those are the only packages whose tests `LD_PRELOAD` the `memory_tools_interpose.so` library. (Those are only packages that mention `LIBRARY_PRELOAD_ENVIRONMENT_VARIABLE` property exported from here)

Locally, if I run the `rcl` tests without this patch on master ros2.repos, I get segfaults in the following tests
```
The following tests FAILED:                                                                                                                                                                                                     
          8 - test_time__rmw_fastrtps_cpp (Failed)                                                                                                                                                                              
         10 - test_context__rmw_fastrtps_cpp (Failed)                                                                                                                                                                           
         17 - test_init__rmw_fastrtps_cpp (Failed)                                                                                                                                                                              
         18 - test_node__rmw_fastrtps_cpp (Failed)                                                                                                                                                                              
         22 - test_guard_condition__rmw_fastrtps_cpp (Failed)                                                                                                                                                                   
         32 - test_time__rmw_fastrtps_dynamic_cpp (Failed)                                                                                                                                                                      
         34 - test_context__rmw_fastrtps_dynamic_cpp (Failed)                                                                                                                                                                   
         41 - test_init__rmw_fastrtps_dynamic_cpp (Failed)                                                                                                                                                                      
         42 - test_node__rmw_fastrtps_dynamic_cpp (Failed)                                                                                                                                                                      
         46 - test_guard_condition__rmw_fastrtps_dynamic_cpp (Failed)                                                                                                                                                           
Errors while running CTest                                                                                                                                                                                                      
--- stderr: rcl                                                                                                                                                                                                                 
Errors while running CTest                                                           
```

But with this patch, 100% of rcl tests pass.

I don't know if the buildfarm is witnessing the same segfaults - it is possible that it is a very subtle combination of variables. It would be nice to see the current state of master compared to this build.

jacobperron commented:
In addition to the current state of master I've also triggered a build with Fast-RTPS 1.7 to see if the time source test failures are coming from something introduced in Fast-RTPS 1.8.

#### Linux CI

* master: [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7079)](https://ci.ros2.org/job/ci_linux/7079/)
* this PR: [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7080)](https://ci.ros2.org/job/ci_linux/7080/)
* this PR + Fast-RTPS 1.7 [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7081)](https://ci.ros2.org/job/ci_linux/7081/)

jacobperron commented:
I guess after https://github.com/ros2/rmw_fastrtps/pull/272, we can't expect Fast-RTPS 1.7 to build.

dirk-thomas commented:
The CI results are a strict improvement over the current master. Merging...
